pages/en/lb4/Creating-components.md

```diff
@@ -307,3 +307,297 @@ class EmailComponent {
+```ts
+// Note that this can also be a class!
+export type Person = {
+  name: string;
+  hobby: string;
+};
+```
+
+### Defining our router and server
+Now that there is a controller to test with, the following code creates a `Router` and `Server`.
+This example uses [express](https://expressjs.com/) and the
+`@types/express` package.
+
+{% include code-caption.html content="/src/server/rpc-router.ts" %}
+```ts
+import {RPCServer} from './rpc-server';
+import * as express from 'express';
+import * as parser from 'body-parser';
+
+export class RPCRouter {
+  routing: express.Router;
+  constructor(private server: RPCServer) {
+    this.routing = express.Router();
+    const jsonParser = parser.json();
+    this.routing.post('*', jsonParser, async (request, response) => {
+      console.log(JSON.stringify(request.body));
+      // Aliasing these values for clarity and brevity.
+      const ctrl = request.body.controller;
+      const method = request.body.method;
+      const input = request.body.input;
+      try {
+        const controller = await this.server.get(`controllers.${ctrl}`);
+        if (!controller) {
+          response.statusCode = 400;
+          response.send(`No controller was found with name "${ctrl}".`);
+          return;
+        }
+        if (!controller[method]) {
+          response.statusCode = 400;
+          response.send(
+            `No method was found on controller "${ctrl}" with name "${method}".`
+          );
+        }
+        response.send(await controller[method](input));
+      } catch (err) {
+        // You'll probably want more robust error handling in a real server!
+        console.error(err);
+        response.statusCode = 500;
+        response.send(err);
+      }
+    });
+    // Use our router!
+    this.server.expressServer.use(this.routing);
+  }
+}
+```
+
+In addition to the server, the example will include a small type definition for
+convenience.
+
+{% include code-caption.html content="/src/server/rpc-server.ts" %}
+```ts
+import {inject, Context} from '@loopback/context';
+import {Server, Application, CoreBindings} from '@loopback/core';
+import {RPCRouter} from './rpc-router';
+import * as express from 'express';
+import * as http from 'http';
+
+export class RPCServer extends Context implements Server {
+  _server: http.Server;
+  expressServer: express.Application;
+  router: RPCRouter;
+  constructor(
+    @inject(CoreBindings.APPLICATION_INSTANCE) public app?: Application,
+    @inject('rpcServer.config') public config?: RPCServerConfig
+  ) {
+    super(app);
+    this.config = config || {};
+  }
+
+  async start(): Promise<void> {
+    this.expressServer = express();
+    this.router = new RPCRouter(this);
+    // Currently, there is a limitation with core node libraries
+    // that prevents the easy promisification of some functions, like
+    // http.Server's listen method.
+    this._server = this.expressServer.listen(
```

Comment:
Can we wrap listen() call with try/catch block?

Reply:
It's meant to be demo code. I don't want to impact its readability with loads of boilerplate safeguarding and error handling.

It's the same reason it's not riddled with debug statements. :P

Reply:
I am actually quite concerned about this too. People look up to our docs for examples they will be copy-n-pasting into their code and which may eventually get into production.

I agree we should keep this readable, but not at the cost of showing bad practices!

Here in particular we should use `p-event` to ensure `error` event reported by the server is correctly handled.

```ts
this._server = this.expressServer.listen(...);
return await pEvent(this._server, 'listen');
```

Reply:
It is also important to wait for `listen` event before returning from `start()`, otherwise our app can print "Server is running on port XYZ" before the server has actually started to listen 😱

What's worse, if there are tests calling `app.start()` and making HTTP requests immediately after `start` returns, then they can sporadically fail when the timing is wrong and `listen` takes more ticks of event-loop/promise micro-queue than usually and the request is started before the server started listening.

Reply:
Including `pEvent` seems a bit odd. I didn't even know it was a thing (I think you mentioned it in your presentation, but I missed the start!). Not everyone will know what pEvent is/does, and I think it detracts from the example, which is why I stuck the comment up there instead.

>It is also important to wait for listen event before returning from start(), otherwise our app can print "Server is running on port XYZ" before the server has actually started to listen 😱

I had a version that handled this using old-school promises:
```ts
return new Promise((resolve, reject) => {
  // using this ugly form to call resolve or reject on the event.
});
```

I wasn't a fan because it detracted from the core logic. I think I'll just use p-event.